💥 full switch to stable middleware (perf reasons)  + ✨  docs website with Nextra

nibtime opened the pull request:
That PR has gotten somewhat big and took ages, as I discovered lots of small changes in Next 12.2 that break things (edge runtime, ISR, Safari ...).

@Shamilik 
thank you very much for reporting the issue with Safari. The problem seems essentially the same as with Firefox as described in #5. Firefox isn't truly CSP-3 compliant and SRI validation gets messed up when used together with `strict-dynamic` (see https://bugzilla.mozilla.org/show_bug.cgi?id=1409200). Presumably, it's the same for Safari, at least for versions 15.4 and 15.5. Once I have confirmation of Safari and Firefox versions that are truly CSP-3 compliant I will update the `strictDynamic` middleware to support them for Hash-based.

So for now, I excluded Safari in this PR from Hash-based strict CSP. Apps won't break now, but unfortunately, also visitors won't be protected by strict CSP if the entry point is a static page. But they will be, if the entry point is a dynamic page - Nonce-based works and will be applied. For static pages, the `fallbackSrc` value of the `strictDynamic` middleware will be used (defaults to `https: 'unsafe-inline'`). I documented that also in the release notes of the upcoming `0.9.0` version.

Another option would be loading Next.js chunks/scripts with an inline proxy script. I already tried that out and that would work with Safari and Firefox as well. Unfortunately, that can break Next.js in a nondeterministic fashion on page reloads and navigation events and I couldn't find a way to resolve that so far.

nibtime committed:
fix(document): pickup `initialProps.head`

suppose its content gets inserted by `next/head` during SSR.

discovered this during docs project setup with nextra

nibtime committed:
fix(middleware): test for file extensions in `isPageRequest` matcher

I think it's agreeable that any path that ends in a dot, is a file and not a page request.
In general, this is safer and more versatile than excluding by base path prefixes

feat: provide logical operators for `ChainMatcher`

`matchNot`, `matchAnd`, `matchOr`

feat: provide matcher for preview mode request (`isPreviewModeRequest`)

nibtime committed:
refactor(middleware): :art: use setup from docs with `/api/reporting` as default of `reporting`

nibtime committed:
docs: :sparkles: set up docs app with nextra

nextra runs on yop of Next 12.2 and this package, so the docs app
is kind of an e2e test itself

docs: shorten package README

docs: setup all-contributors

chore: symlink from apps/docs/pages/CONTRIBUTING.mdx to ./CONTRIBUTING.md

chore: symlink from packages/next-safe-middleware/LICENSE to ./LICENSE

build(deps): add all-contributors-cli

nibtime committed:
build: update build scripts fur mutliple apps

nibtime committed:
feat(middleware): :boom: new `MiddlewareChainContext` interface

perf: do stuff for reduced middleware CPU utilization
- use `ctx.get` and `ctx.set` for caching CSP in for middleware chain (no need to serialize/deserialize CSP)
- write to repsonse only once from chain cache with a finalizer
- remove unnecessary double ops like `uniq`

refactor: modularize code and types in `compose` and `builder` barrels

feat: extra bundle just for composition: `dist/compose`

docs: fix some mistakes in JSDoc

BREAKING CHANGE: turn positional params into named params  for `Configinitializer`

BREAKING CHANGE: drop ua-parser-js for built-in `userAgent` from `next/server` (since 12.2)

BREAKING CHANGE: `ChainableMiddleware` with `(ctx: `MiddlewareChainContext)` as 3rd parameter.

Affects only middleware building and composition, not security middleware
Those features haven't been really documented so far, so probably no one will notice

nibtime committed:
fix(csp): handle boolean directives correctly

refactor(utils): avoid unnecessary ops

test: module test for boolean directives

nibtime committed:
chore(changesets): :heavy_plus_sign: add

nibtime committed:
build(deps): only next >=12.2 supported now (open-ended peer dependency)

because of focus on Stable middleware

remove @types/ua-parser-js (and one other)

remove @types/base-64 (and one other)

nibtime committed:
fix(document): :bug: use children?: any for `provideComponents`

fixes #46. In React 18 typings, FC doesn't have auto children anymore

nibtime committed:
:busts_in_silhouette: Add @nibtime as a contributor

nibtime committed:
build(deps): :arrow_up: build(deps): upgrade workspace to React 18 and Next 12.2.2

And set children and return type to `any` for `provideComponents`.
For some weird type reason keeps failing on Vercel , but not in GitPod.
fixes #46 (and drove me nuts)

nibtime committed:
fix(e2e): some React 18 error in _app

nibtime committed:
build(test): upgrade workspace to jest 28

fix(test): temporarily disable React tests:  act(...) is not supported in production builds of React.

see https://github.com/vercel/next.js/discussions/35676

Shamilik commented:
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/2459175/179313308-867e2c2b-1c8a-41fc-bfd8-3df07378d1c7.png">
Hello. Preview isn't working on Safari 15.5.

Received CSP:
```
Content-Security-Policy: default-src 'self';object-src 'none';base-uri 'none';img-src 'self' data: https://vercel.com https://img.shields.io https://*.githubusercontent.com https://gitpod.io;font-src 'self' https://rsms.me;style-src 'self' https://rsms.me 'sha256-Sr0R23aPEYkZEIraULoANv0+lJ6nm++YjMKNjJJgUXA=' 'sha256-lgodCpuy1TR56k3x67FV95N+KM+XG8BOqKMMa/XHk28=' 'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=' 'sha256-GHgVb5EEAUPhVeTfwEFuX9+WADcCrePlK6CwdXznIhY=' 'sha256-xd6kVmsB5qDY47QB4x7Ys+1t88OSTa40q/98d9NyF8Q=' 'sha256-8BNxsIsc6VHj8/elC63fqbrGsnTOvhNTf17uhaIdUI4=' 'sha256-d72pVhmRTNXT2Gr2OHFRLnVaHBfiBI5EvDCF6tA924Y=' 'sha256-wVdJrANUOK6FSKXAK/b1cyEUJMPQfQmGgKD56BosKsQ=' 'sha256-QuMjhq9r7jLn8e9zqxh8mfMbl0arL7LCu/hyoLaFBQc=' 'sha256-9HGruJg4WccHXas5I1NmLn7tI1TDh6N26o6+/dy8sm4=' 'sha256-WhdL3NGEYeeQu4IJi8MIwu7bKqNwzqfyf0ZxdTEe46s=' 'sha256-QvCvI4erjhH7DNwV83hMvuRvdfRtk16q9GHz5wlS0bs=' 'sha256-jKE6QZqne5OsrfemNvuLSNoud++NsCOiSlGuIsQns5o=' 'sha256-J/RYDozRaa3WUwPsZ82tQdcUhZyfAGMmyxZ1ILqy40k=' 'sha256-/cz+p719dOFygDAqDgEjHhHSRaka+kWXk3WHAOXiURk=' 'sha256-k1LZnPBgTDfZ8Oz4iDbO2KHJG4Kq7uqZ7tj4buN8+j4=' 'sha256-OpF46MbWec8lEJE9VZEkMMzTCPNuJOw+MYd0C2lV6eQ=' 'sha256-cXE5WSWHJNr6CysOfi+H1lA3XpTfgZFzu2+5D8P/j6M=' 'sha256-ioeco5QPBMU6V5vUrY+ASxoxaQnK7/gRpV0D5ZCKpK4=' 'unsafe-hashes' 'report-sample';script-src 'strict-dynamic' https: 'unsafe-inline' 'sha256-FRcCAqWeT9c2BDWISR4nAeoh5T3cbINwn0j4tzAs9PU=' 'sha256-2AH2VH1B+5jBCXgwJkVuj1TRGO0aUm9dcNatWrnyosU=' 'sha256-lBDbadDkyeD7bOL6E1HDHFZXy+jFNq5d5QL2yspKv/A=' 'sha256-9KBZhMnIhr6i6Vtn/+m7sBLmWhnSfquC+BYUfYZs21E=' 'sha256-+n6qCniDL16DO3FYEyaFapnzcUr0xSMx30D68AgajH4=' 'sha256-AClCwzlXIejFvVduN+HneWJr65UFQ5kUOkgPapcZT2w=' 'sha256-by3fHBuyRxdqAMf7fSVCgE6zVnZIa/6qP5bmgf2pd9g=' 'sha256-l6DHoC1Ab+rmjDBGLFym9joywrc+95/rgKGChtq9AcE=' 'sha256-FCZEmVkjYzI1TqZIyDWAwXAH3kbZnFdlIejPw8aqSbk=' 'sha256-6OOA1hrdgdwg6KGy93+51o0591+S8gPkMcSh0mfrnRc=' 'sha256-3Tbstl/7NhKqVtTQ8qik7oOw8djYptFHzQZjgEcFI/8=' 'sha256-R5PHcVi5p/1cmbT7VOzRSRwi9bwDn60L3kQ7v41eXnM=' 'sha256-a5Q14E9Z/8Rc/lZYyEjEbIN1uUR9bs+hGCGl+iuFC0o=' 'sha256-0PsJr+/QmZoW1v5o0JdOePeFfi1HUspiKkFAucTNWNM=' 'sha256-gsOvp62vNJoRqsvQWJs67x3+lvazIWWDBNgg6ybMRmg=' 'sha256-kEM9zyBXuBR3jn6hhfNopeBCB4H3SCiZMFYK1THK6Nc=' 'sha256-Mm1eOSIv97qBO9IsiSracSPKT4mcIjBOepVpnr6GxXY=' 'sha256-foBhF47FLjBw9yVL8WE2GQtf7M2av4UqyUSVXsMw94Q=' 'sha256-B5Nj5tsbu/Mz0wqhtzuQ9wSbxr1hzxe23gr5v+6Yy6Q=' 'sha256-/mgaywVXJ1yHET9ftr9ak3qzzHBVG9IfuG0llKYbKJE=' 'sha256-kxviE/tyzwqZqaavwhRUYmXttKyTtPz0wpp4Pp/tfUY=' 'sha256-nXzjemf1oMXVn/mFfKiTPCCp8xPTf/EehdjiyD1Vag0=' 'sha256-2J4eFP5LdsUNsymmd+5AMuJ56gwkZ1DtwRbeJT9Qtp4=' 'sha256-VbFiIqooxXDId2fGnqVP5Tvmx+1mUVChJfS95XOfe1I=' 'sha256-ROdpclxijlXXVBWqulEKgL8WHbSrEtJpJ4/Ah1Vk400=' 'sha256-zEkrGm56GbCqXjLxIM90BBjkU8SOsC32PWVQfGOovZ0=' 'sha256-mGP5vHbto9iZfd/ytJdBtFZ0KMePO5x1snKLxDOip7s=' 'sha256-o5xkS8gmT76hLNAHYl2Invs2VKT5kWot+ILoRYbDmsM=' 'sha256-2rNZhgVKOKupmxSfMhy8NXjyiFu5JwZX8ijD2lfA+Uw=' 'sha256-/mw7+IBTDgIBr7RcI2lZxV77k0HVCwt2bwSNvlqODqQ=' 'report-sample';report-uri /api/reporting;report-to default
```

nibtime committed:
perf(middleware): :zap: some adaptions to reduce execution time/utilization

nibtime committed:
perf(middleware): provide `telemetry` middleware

nibtime committed:
fix(apps): update for Next 12.2, React 18 and breaking changes

perf: add opt-in middleware performance logging to apps

fix(docs): custom static 404 page

docs: update docs site for breaking changes

docs: update README for breaking changes

docs: add guide for tracing middleware

update @builder.io/partytown to ^0.6.4 (and one other)

nibtime committed:
fix(middleware): :bug: get rid of some bugs

fix: consider basePath for fetching hashes

fix: prepend basePath to relative URLs of reporting middleware

fix: matcher to exclude `x-nextjs-data` requests

fix: exclude Safari in `supportsSrcIntegrityCheck`

refactor: move deepFreeze to utils

nibtime committed:
chore(changesets): add/update

nibtime committed:
fix(document): :bug: get rid of bugs introduced by Next 12.2

fix: don't apply nonce to ISR requests
- detection of difference to gssp has become impossible
- new `gsspWithNonceAppliedToCsp` wrapper must be used in every route with `getServerSideProps`

fix: consider basePath when rewriting URLs to paths
- relevant for `nextScriptWithInjectedIntegrity`, otherwise can't find them in file system and inject integrity

fix: hash new `beforeInteractiveInlineScripts` of <Head>

fix: adapt noncing hashing for `experimental-edge` runtime
- with `experimental-edge` runtime, scripts are loaded differently
- `<head>` holds preload links and end of body loads scripts asynchronously
- inject integrity into preload links in `Head` components for Hash-based routes
- remove integrity from preload links in `Head` components for Nonce-based routes
- apply hashing/noncing in provided `NextScript` component (end of body)

fix: remove integrity from scripts when nonce available
- else preloading scripts will fail as they have no integrity

fix: update nonce application on preNextScripts
- for some reason, partytown inline scripts didn't get nonced anymore
- partytown inline scripts also don't get nonced by Next when left alone

fix: only apply nonce in prod builds

refactor: `logCtxHeaders` to conveniently log ctx headers of `_document`

BREAKING CHANGE: CSP nonce is no longer added automatically added to dynamic routes with `getServerSideProps`

BREAKING CHANGE: class export of already deprecated `NextSafeDocument` had to be dropped due to changes because of module conflicts

BREAKING CHANGE: drop `enhanceAppWithNonce`, no longer needed. Avoiding it is also good, as customizing `renderPage` is discouraged: https://nextjs.org/docs/advanced-features/custom-document#customizing-renderpage

nibtime commented:
@all-contributors add @Shamilik for bug
